daemon/cmd/daemon.go

```diff
@@ -582,8 +582,76 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	debug.RegisterStatusObject("ongoing-endpoint-creations", d.endpointCreations)
 
 	d.k8sWatcher.RunK8sServiceHandler()
+
+	if option.Config.DNSPolicyUnloadOnShutdown {
+		log.Debug("Registering cleanup function to unload DNS policies")
+
+		// add to pre-cleanup funcs because this needs to run on graceful shutdown, but
+		// before the relevant subystems are being shut down.
+		cleaner.preCleanupFuncs.Add(func() {
+			// Stop k8s watchers
+			log.Info("Stopping k8s service handler")
+			d.k8sWatcher.StopK8sServiceHandler()
+
+			// Iterate over the policy repository and remove L7 DNS part
+			removeL7DNSRules := func(pr policyAPI.Ports) error {
+				portProtocols := pr.GetPortProtocols()
+				if len(portProtocols) != 1 || portProtocols[0].Port != "53" {
+					// not a L7 DNS rule
+					return nil
+				}
+				portRule := pr.GetPortRule()
+				if portRule == nil || portRule.Rules == nil {
+					// no L7 rule
+					return nil
+				}
+				l7Rules := portRule.Rules
+				log.Debugf("Found egress L7 DNS rule, portProtocol %#v l7 rules %#v", portProtocols[0], *l7Rules)
+
+				// For security reasons, the L7 DNS policy must be a
+				// wildcard in order to trigger this logic.
+				// Otherwise we could invalidate the L7 security
+				// rules. This means if any of the DNS L7 rules
+				// have a matchPattern of * then it is OK to delete
+				// the L7 portion of those rules.
+				hasWildcard := false
+				for _, dns := range l7Rules.DNS {
+					if dns.MatchPattern == "*" {
+						hasWildcard = true
+						break
+					}
+				}
+				if hasWildcard {
+					portRule.Rules = nil
+				}
+				return nil
+			}
+
+			policyRepo := d.GetPolicyRepository()
+			policyRepo.Mutex.Lock()
+			policyRepo.IterateLocked(func(rule *policyAPI.Rule) {
+				for _, er := range rule.Egress {
+					_ = er.ToPorts.Iterate(removeL7DNSRules)
+				}
+				// TODO: modify Ingress, IngressDeny and EgressDeny as well?
+			})
+			policyRepo.Mutex.Unlock()
+
+			// Bump revision to trigger policy recalculation
+			log.Info("Triggering policy recalculation")
+			policyRepo.BumpRevision()
+			regenerationMetadata := &regeneration.ExternalRegenerationMetadata{
+				Reason:            "unloading DNS policies on graceful shutdown",
+				RegenerationLevel: regeneration.RegenerateWithDatapathRewrite,
```

Comment:
I believe that `RegenerateWithoutDatapath` should be sufficient for this use case, as "reload" and "rebuild" of the datapath is not required. I think that the regeneration levels are not sufficiently documented to justify this, but my understanding is that `RegenerateWithoutDatapath` will do policy recalculation + bpf policymap population. Lowering this value will allow us to skip some of the more expensive steps like compiling BPF programs or loading BPF programs into the kernel (which should not be necessary to implement this desired change).

Worth double-checking; worst case we _can_ use a higher regeneration level like this, it's just more expensive (possibly adding seconds onto shutdown time). But these regeneration levels are intended to allow policy regeneration like this without some of those more expensive steps.

Reply:
Changed it to `RegenerateWithoutDatapath` and the effect is still the same in my tests.

I think I must have blindly copied `regenerationMetadata` with `RegenerateWithDatapathReload` from somewhere else and forgot to review/change its field values :facepalm: